diff --git a/.bazelversion b/.bazelversion
index 0b2eb36f508..ee74734aa22 100644
--- a/.bazelversion
+++ b/.bazelversion
@@ -1 +1 @@
-3.7.2
+4.1.0
diff --git a/tensorflow/lite/java/gpu_version_script.lds b/tensorflow/lite/java/gpu_version_script.lds
index f298833e174..559367b5d1e 100644
--- a/tensorflow/lite/java/gpu_version_script.lds
+++ b/tensorflow/lite/java/gpu_version_script.lds
@@ -5,6 +5,15 @@ VERS_1.0 {
     JNI_OnLoad;
     JNI_OnUnload;
     TfLiteGpu*;
+    TfLite*;
+    *TfLite*;
+    *tflite*;
+    *TFL_*;
+    *FlatBufferModel*;
+    *BuiltinOpResolver*;
+    *Interpreter*;
+    *MutableOpResolver*;
+    *ErrorReporter*;
 
   # Hide everything else.
   local:
diff --git a/tensorflow/lite/java/tflite_version_script.lds b/tensorflow/lite/java/tflite_version_script.lds
index 9634a802356..f16d7e78c95 100644
--- a/tensorflow/lite/java/tflite_version_script.lds
+++ b/tensorflow/lite/java/tflite_version_script.lds
@@ -5,7 +5,15 @@ VERS_1.0 {
     JNI_OnLoad;
     JNI_OnUnload;
     TfLite*;
-
+    *TfLite*;
+    *tflite*;
+    *TFL_*;
+    *FlatBufferModel*;
+    *BuiltinOpResolver*;
+    *Interpreter*;
+    *MutableOpResolver*;
+    *ErrorReporter*;
+    
   # Required for libunwind. This is needed if built and then run internally.
     google_find_phdr;
 
